archivebot.py

```diff
@@ -1,178 +1,211 @@
-        else:
-            send_message('No results found', event['channel'])
-    except ValueError as e:
-        print(traceback.format_exc())
-        send_message(str(e), event['channel'])
-
-def handle_message(event):
-    if 'text' not in event:
-        return
-    if 'username' in event and event['username'] == 'bot':
-        return
-
-    try:
-        print(event)
-    except:
-        print("*"*20)
-
-    # If it's a DM, treat it as a search query
-    if event['channel'][0] == 'D':
-        handle_query(event)
-    else: # Otherwise save the message to the archive.
-        cursor.executemany('INSERT INTO messages VALUES(?, ?, ?, ?)',
-            [(event['text'], event['user'], event['channel'], event['ts'])]
-        )
-        conn.commit()
-
-# Loop
-if sc.rtm_connect():
-    while True:
-        try:
-            for event in sc.rtm_read():
-                if event['type'] == 'message':
-                    handle_message(event)
-        except WebSocketConnectionClosedException:
-            sc.rtm_connect()
-        except:
-            print(traceback.format_exc())
-        time.sleep(1)
-else:
-    print("Connection Failed, invalid token?")
+import datetime
+import os
+import sqlite3
+import time
+import traceback
+
+from slackclient import SlackClient
+from websocket import WebSocketConnectionClosedException
+
+# Connects to the previously created SQL database
+conn = sqlite3.connect('slack.sqlite')
+cursor = conn.cursor()
+cursor.execute('create table if not exists messages (message text, user text, channel text, timestamp text, UNIQUE(channel, timestamp) ON CONFLICT REPLACE)')
+cursor.execute('create table if not exists users (name text, id text, text avatar, UNIQUE(id) ON CONFLICT REPLACE)')
+cursor.execute('create table if not exists channels (name text, id text, UNIQUE(id) ON CONFLICT REPLACE)')
+
+# This token is given when the bot is started in terminal
+slack_token = os.environ["SLACK_API_TOKEN"]
+
+# Makes bot user active on Slack
+# NOTE: terminal must be running for the bot to continue
+sc = SlackClient(slack_token)
+
+# Double naming for better search functionality
+# Keys are both the name and unique ID where needed
+ENV = {
+    'user_id': {},
+    'id_user': {},
+    'channel_id': {},
+    'id_channel': {}
+}
+
+# Uses slack API to get most recent user list
+# Necessary for User ID correlation
+def update_users():
+    info = sc.api_call('users.list')
+    ENV['user_id'] = dict([(m['name'], m['id']) for m in info['members']])
+    ENV['id_user'] = dict([(m['id'], m['name']) for m in info['members']])
+
+    args = []
+    for m in info['members']:
+        args.append((
+            m['name'],
+            m['id'],
+            m['profile'].get('image_32', 'https://secure.gravatar.com/avatar/c3a07fba0c4787b0ef1d417838eae9c5.jpg?s=32&d=https%3A%2F%2Ffst.slack-edge.com%2F66f9%2Fimg%2Favatars%2Fava_0024-32.png')
+        ))
+    cursor.executemany("INSERT INTO users(name, id, avatar) VALUES(?,?,?)", args)
```

Comment:
I get an error here due to the create table command having `text avatar` instead of `avatar text`.